### PR TITLE
adapter: improve routing hints

### PR DESCRIPTION
**Pull request by mjibson**
Improve an error message about auto routing.

See https://github.com/MaterializeInc/materialize/issues/24718#issuecomment-1912342691

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - n/a

### CODE REVIEW

File: src/adapter/src/error.rs [hunk 1]
```diff
@@ -302,6 +303,9 @@ impl AdapterError {
                     earliest_possible,
                 ))
             }
+            AdapterError::UnallowedOnCluster { cluster, .. } => (cluster == MZ_INTROSPECTION_CLUSTER.name).then(||
+                format!("The transaction is executing on the {cluster} cluster, maybe having been routed there by the first statement in the transaction.")
```
Comment (mjibson) [inline src/adapter/src/error.rs:307]: Is this too wordy? I'm not sure how to shorten it usefully.

Comment (jkosh44) [reply src/adapter/src/error.rs:307]: I think it's fine.

**Review by jkosh44 — APPROVED** [review]
